documentation/commandline_options.md
## Supported Command Line Options

### 1. Listening Host Configuration

Optional. By default, Azurite V3 will listen to 127.0.0.1 as a local server.
You can customize the listening address per your requirements.

#### Only Accept Requests in Local Machine

```cmd
--blobHost 127.0.0.1
--queueHost 127.0.0.1
--tableHost 127.0.0.1
```

#### Allow Accepting Requests from Remote (potentially unsafe)

```cmd
--blobHost 0.0.0.0
--queueHost 0.0.0.0
--tableHost 0.0.0.0
```

### 2. Listening Port Configuration [Optional]

By default, Azurite V3 will listen to 10000 as blob service port, and 10001 as queue service port, and 10002 as the table service port.
You can customize the listening port per your requirements.

> Warning: After using a customized port, you need to update connection string or configurations correspondingly in your Storage Tools or SDKs.
> If starting Azurite you see error `Error: listen EACCES 0.0.0.0:10000` the TCP port is most likely already occupied by another process.

#### Customize Blob/Queue Service Listening Port

```cmd
--blobPort 8888
--queuePort 9999
--tablePort 11111
```

#### Let System Auto Select an Available Port

```cmd
--blobPort 0
--queuePort 0
--tablePort 0
```

> Note: The port in use is displayed on Azurite startup.

### 3. Workspace Path Configuration [Optional]

Azurite V3 needs to persist metadata and binary data to local disk during execution.

You can provide a customized path as the workspace location, or by default, Current process working directory will be used.

```cmd
-l c:\azurite
--location c:\azurite
```

### 4. Access Log Configuration [Optional]

By default Azurite will display access log in console. **Disable** it by:

```cmd
-s
--silent
```

### 5. Debug Log Configuration [Optional]

Debug log includes detailed information on every request and exception stack traces.
Enable it by providing a valid local file path for the debug log destination.

```cmd
-d path/debug.log
--debug path/debug.log
```

### 6. Loose Mode Configuration [Optional]

By default Azurite will apply strict mode. Strict mode will block unsupported request headers or parameters. **Disable** it by enabling loose mode:

```cmd
-L
--loose
```

### 7. Certificate Configuration (HTTPS) [Optional]

By default Azurite will listen on HTTP protocol. Provide a PEM or PFX certificate file path to enable HTTPS mode:

```cmd
--cert path/server.pem
```

When `--cert` is provided for a PEM file, must provide coresponding `--key`.

```cmd
--key path/key.pem
```

When `--cert` is provided for a PFX file, must provide coresponding `--pwd`

```cmd
--pwd pfxpassword
```

### 8. OAuth Configuration [Optional]

By default, Azurite doesn't support OAuth and bearer token. Enable OAuth authentication for Azurite by:

```
--oauth basic
```

> Note. OAuth requires HTTPS endpoint. Make sure HTTPS is enabled by providing `--cert` parameter along with `--oauth` parameter.

Currently, Azurite supports following OAuth authentication levels:

#### Basic

In basic level, `--oauth basic`, Azurite will do basic authentication, like validating incoming bearer token, checking issuer, audience, expiry. But Azurite will NOT check token signature and permission.

### 9. Skip API Version Check [Optional]

By default Azurite will check the request API version is valid API version. Skip the API version check by:

```cmd
--skipApiVersionCheck
```

### 10. Disable Product Style Url [Optional]

When using FQDN instead of IP in request Uri host, by default Azurite will parse storage account name from request Uri host. Force parsing storage account name from request Uri path by:

```cmd
--disableProductStyleUrl
```

### 11. Command Line Options Differences between Azurite V2

Azurite V3 supports SharedKey, Account Shared Access Signature (SAS), Service SAS, OAuth, and Public Container Access authentications, you can use any Azure Storage SDKs or tools like Storage Explorer to connect Azurite V3 with any authentication strategy.

An option to bypass authentication is **NOT** provided in Azurite V3.


documentation/environment_variable_options.md
## Supported Environment Variable Options

When starting Azurite from npm command line `azurite` or docker image, following environment variables are supported for advanced customization.

### Customized Storage Accounts & Keys

Azurite V3 allows customizing storage account names and keys by providing environment variable `AZURITE_ACCOUNTS` with format `account1:key1[:key2];account2:key1[:key2];...`.

For example, customize one storage account which has only one key:

```cmd
set AZURITE_ACCOUNTS="account1:key1"
```

Or customize multi storage accounts and each has 2 keys:

```cmd
set AZURITE_ACCOUNTS="account1:key1:key2;account2:key1:key2"
```

Azurite will refresh customized account name and key from environment variable every minute by default. With this feature, we can dynamically rotate account key, or add new storage accounts on the air without restarting Azurite instance.

> Note. Default storage account `devstoreaccount1` will be disabled when providing customized storage accounts.

> Note. Should update connection string accordingly if using customized account name and key.

> Note. Use `export` keyword to set environment variable in Linux like environment, `set` in Windows.

### Customized Metadata Storage by External Database (Preview)

By default, Azurite leverages [loki](https://github.com/techfort/LokiJS) as metadata database.
However, as an in-memory database, loki limits Azurite's scalability and data persistency.
Set environment variable `AZURITE_DB=dialect://[username][:password][@]host:port/database` to make Azurite blob service switch to a SQL database based metadata storage, like MySql, SqlServer.

For example, connect to MySql or SqlServer by set environment variables:

```bash
set AZURITE_DB=mysql://username:password@localhost:3306/azurite_blob
set AZURITE_DB=mssql://username:password@localhost:1024/azurite_blob
```

When Azurite starts with above environment variable, it connects to the configured database, and creates tables if not exist.
This feature is in preview, when Azurite changes database table schema, you need to drop existing tables and let Azurite regenerate database tables.

> Note. Need to manually create database before starting Azurite instance.

> Note. Blob Copy & Page Blob are not supported by SQL based metadata implementation.

> Tips. Create database instance quickly with docker, for example `docker run --name mysql -p 3306:3306 -e MYSQL_ROOT_PASSWORD=my-secret-pw -d mysql:latest`. Grant external access and create database `azurite_blob` using `docker exec mysql mysql -u root -pmy-secret-pw -e "GRANT ALL PRIVILEGES ON *.* TO 'root'@'%' WITH GRANT OPTION; FLUSH PRIVILEGES; create database azurite_blob;"`. Notice that, above commands are examples, you need to carefully define the access permissions in your production environment.


ChangeLog.md
# Changelog

> Note. This file includes changes after 3.0.0-preview. For legacy Azurite changes, please goto GitHub [releases](https://github.com/Azure/Azurite/releases).

## Upcoming Release

General:

- Updated Readme by adding account key must be base64 encoded string.

Table:

- Correctly responds with status 202 on merge with non-existent entity.
- Properly differentiate between upsert and update in batch merge and replace.
- Added additional tests via raw REST tests.
- Correctly deletes a table that is a substring of another table.

## 2022.06 Version 3.18.0

General:

- Bump up service API version to 2021-08-06
- Modified the error messge for invalid API version to make it more actionable.

Blob:

- Fixed issue that startCopyFromURL and copyFromURL API not fail, when request container if-none-match="\*" and dest blob already exist.

Table:

- Reject table batch request bodies exceeding 4MB.
- Fix binary table property validation to be 64K bytes not 32K characters.
- Does not error when table created is a substring of another table.
- Correctly responds with status 404 on patch with non-existant entity.
- Fix pagination when no rowkey in continuation token

## 2022.04 Version 3.17.1

Table:

- Removes commas from RegEx checking key validity.
- Updated property check to handle null property and added regression test.

## 2022.04 Version 3.17.0

General:

- Bump up service API version to 2021-06-08
- Fixed SAS validation failure for version 2020-12-06 and later

Table:

- Fixed empty partition key and row key handling in batch write operations.
- Fixed batch reponse for Go SDK, includes additional CRLF on closure of changesetresponse section.
- Removed query strings from Location and DataServiceId batch response headers.
- Modified the deserialization of batch request for case that a raw / not url encoded % is present in the body.
- Added additional tests and checks for table names on creation.
- Added more granularity and precision to etags.
- Added checks for invalid characters in partition and row keys.
- Rejects entities with string props longer than 32K chars.
- Added check for body length greater than 4MB.

## 2022.02 Version 3.16.0

General:

- Bump up service API version to 2021-04-10
- Ensure the storage location exists, and allow relative paths in the VSCode extension settings that are resolved based on the workspace folder.
- Update Azure CI to use latest image of windows due to deprecation of `vs2017-win2016` image

Blob:

- Fixed issue that startCopyFromURL and copyFromURL API not respect `--disableProductStyleUrl` parameter in parse source Uri.

Queue:

- Fixed issue that queue list result is not in alphabetical order.
- Fixed class name of QueueSASAuthenticator mistakenly named BlobSASAuthenticator.

Table:

- Fixed issues with deleting entities using empty string for RowKey.
- Fixed HTTP 500 causes by continuation token containing non-ASCII. Values are now encoded with base64.
- Fixed a table sas test case failure.
- Added support for batch transaction rollback on error in batch.
- Fixes issues with Whitespacing in Table Queries
- Fixes issue with Edm Type Validation
- Fixes issue when trying to add entity with Boolean or Int32
- Failed table transaction correctly returns 409 Status code
- Refactors tests for Table APIs
- Adds several tests for Table APIs
- Fixes issues for upsert and merge with etag matching
- Allow any valid weak etag even though we know it will fail with a 412
- Added check for table query validity

## 2021.12 Version 3.15.0

General:

- Bump up service API version to 2021-02-12
- Fixed access to secondary location with IP style Uri from JS/.net SDK failure.
- Fixed an issue in Visual Studio Code extension, by changing the Location with relative path, from base on Visual Studio Code installation path, to base on the current opened workspace folder.

Blob:

- Fixed start copy blob fail with `x-ms-access-tier` header and from Archive blob in same account.

## 2021.10 Version 3.14.3

General:

- Added new parameter `--disableProductStyleUrl`, to force parsing storage account from request Uri path, instead of from request Uri host.
- Restored ability to connect to host.docker.internal.

Blob:

- Fixed list blob API "include" query parameter not work when not lower case, by make it case insensitive.
- Supported list container/blob with "include" query parameter as empty string.
- Added more allowed value to list blob request "include" query parameter:'tags', 'versions', 'deletedwithversions', 'immutabilitypolicy', 'legalhold', 'permissions'.
- Added more allowed value to list container request "include" query parameter: 'deleted'.
- Raised 416 when start range is bigger than blob length.
- Fixed issue that duplicated decode rscd, rsce, rscl and rsct of SAS token in input request Uri.

Queue:

- Fixed issue that expired message still can be get, peek, update, delete.

Table:

- Supported basic level of OAuth autentication on Table service.
- Removed extra CRLF from batch transaction response which caused issues for Microsoft.Azure.Cosmos.Table NuGet package.

Table:

- Fixed issue with incorrect results returned when using boolean values in query.
- Fixed issue with incorrect results returned with whitespacing and parens with int64 values in query.

## 2021.9 Version 3.14.2

Blob:

- Supported rscc, rscd, rsce, rscl, rsct query parameters in SAS tokens.
- Fixed Blob_Download API by adding header `x-ms-creation-time` in responds.

Table:

- Added getServiceProperties response.
- Added setServiceProperties response.
- Fixed paged queries across partitions.

## 2021.7 Version 3.14.1

General:

- Added support for generating standalone azurite.exe.

Table:

- Correctly returning the results of paginated queries.
- Added filter support for Query Tables operation.
- Corrected tokenization of queries in table storage.

## 2021.7 Version 3.14.0

General:

- Bump up service API version to 2020-10-02

lib/core/env.ts
'use strict';

import path from 'path';
import BbPromise from 'bluebird';
import crypto from 'crypto';
import fsextra from 'fs-extra';

const fs: any = BbPromise.promisifyAll(fsextra);

let initialized = false;

class Environment {
    silent: any;
    azuriteWorkspacePath: any;
    azuriteRootPath: string;
    accountAuth: any;
    dbNameBlob: string;
    dbNameTable: string;
    localStoragePath: string;
    azuriteDBPathBlob: string;
    azuriteDBPathTable: string;
    emulatedStorageAccountName: string;
    blobStoragePort: any;
    queueStoragePort: any;
    blobModulePath: string;
    queueModulePath: string;
    tableModulePath: string;
    tableStoragePort(tableStoragePort: any, arg1: () => void): any {
        throw new Error("Method not implemented.");
    }
    constructor() {
    }

    init(options) {
        if (initialized && !options.overwrite) {
            return BbPromise.resolve();
        }
        initialized = true;
        this.azuriteWorkspacePath = options.l || options.location || process.cwd();
        this.azuriteRootPath = path.join(__dirname, '../..');
        this.silent = options.s || options.silent;
        this.accountAuth = options.a || options.accountAuth;
        this.dbNameBlob = '__azurite_db_blob__.json';
        this.dbNameTable = '__azurite_db_table__.json';
        this.localStoragePath = path.join(this.azuriteWorkspacePath, '__blobstorage__');
        this.azuriteDBPathBlob = path.join(this.azuriteWorkspacePath, this.dbNameBlob);
        this.azuriteDBPathTable = path.join(this.azuriteWorkspacePath, this.dbNameTable);
        this.emulatedStorageAccountName = 'devstoreaccount1';
        this.blobStoragePort = options.p || options.blobPort || 10000;
        this.queueStoragePort = options.q || options.queuePort || 10001;
        this.tableStoragePort = options.t || options.tablePort || 10002;
        this.blobModulePath = path.join(this.azuriteRootPath, 'bin', 'blob');
        this.queueModulePath = path.join(this.azuriteRootPath, 'bin', 'queue');
        this.tableModulePath = path.join(this.azuriteRootPath, 'bin', 'table');
        return fs.mkdirsAsync(this.localStoragePath);
    }

    /**
     * Based on the request it creates the according URI that is served by Azurite's internal web interface 
     * directly powered by Node's static file server.
     * 
     * The id is hashed to avoid base64-encoded filenames (i.e. ids) longer than 255 characters which is not supported on some file systems.  
     * 
     * @param {string} id of the blob
     * 
     * @memberof Environment
    * */
    webStorageUri(id) {
        const hash = crypto.createHash('sha1').update(id).digest('base64').replace(/\//g, '_');
        return `http://localhost:${this.blobStoragePort}/blobs/${hash}`;
    }

    /**
     * Creates the full path to the location of a blob on disk based on its ID. 
     * 
     * @param {any} id 
     * @returns full path to blob on disk
     * @memberof Environment
     */
    diskStorageUri(id) {
        const hash = crypto.createHash('sha1').update(id).digest('base64').replace(/\//g, '_');
        return path.join(this.localStoragePath, hash);
    }

    // We prepend a specific character to guarantee unique ids.
    // This is neccessary since otherwise snapshot IDs could overlap with block IDs could overlap with block-/append-/page-blob IDs.
    blobId(containerName, blobName) {
        return Buffer.from(`A${containerName}${blobName}`, 'utf8').toString('base64');
    }

    blockId(containerName, blobName, blockId) {
        return Buffer.from(`B${containerName}${blobName}${blockId}`, 'utf8').toString('base64');
    }

    snapshotId(containerName, blobName, date) {
        return Buffer.from(`C${containerName}${blobName}${date}`, 'utf8').toString('base64');
    }
}

export default new Environment();
